Refactor Cilium agents and operator to depend on kvstore cell

giorio94 opened the pull request:
Please review commit by commit, and refer to the individual descriptions for additional information.

Sorry for the large PR. Unfortunately, splitting it further is tricky because, although all commits are mostly self-contained and are expected to compile and pass the tests correctly, a few intermediate states are characterized by subtle quirks potentially causing problems in certain conditions.

giorio94 committed:
kvstore: rename Client() to LegacyClient()

In preparation for the subsequent introduction of the kvstore.Client
interface, let's rename the legacy global accessor, which is intended
to be removed after migrating all users to the kvstore Cell.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: introduce Client interface

Introduce the kvstore.Client interface, which mimics the k8s.Clientset
one. This interface is intended to be provided via Hive and replace the
current BackendOperations promise, ensuring better usability and a more
predictable lifecycle (i.e., the client is guaranteed to be initialized
once the corresponding start hook terminates). The stop hook now also
takes care of actually closing the kvstore client. Indeed, that logic
no longer revokes the leases, but only takes care of closing all
connections.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: switch SetupDummy to return kvstore.Client

In preparation for the subsequent migration, let's switch the functions
setting up the kvstore client for testing purposes to return the
kvstore.Client interface. Additionally, let's add the possibility of
returning a disabled client when specifying the empty backend.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: let kvstore.Cell provide kvstore.Client, and adapt users

Update the kvstore.Cell to provide the newly introduced Client
interface, rather than a BackendOperations promise. This allows
simplifying the current users thanks to a more straightforward
lifecycle.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore/heartbeat: rewrite using a timer job

Rewrite the kvstore heartbeat cell to make use of a timer job, rather
than bespoke logic, to make use of the built-in observability features.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore/cell: make default kvstore backend configurable

Let's make the default backend configurable in preparation for
migrating Cilium agents and operator to leverage the kvstore
cell. Indeed, by default, the kvstore client is disabled in
these components (i.e., the backend is set to the empty string),
while it is set to etcd in the clustermesh-apiserver.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
daemon: depend on the kvstore cell

Add the kvstore cell to the list of the agent dependencies, and get rid
of the bespoke initialization logic. The start hook of the kvstore cell
is modified to still set the global variables, to allow for a gradual
migration of the remaining accesses over the next commits.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
operator: depend on the kvstore and heartbeat cells

Add the kvstore cell to the list of operator dependencies, and get rid
of the bespoke initialization logic. As a first user, let's also
register the already existing heartbeat cell for consistency.

The kvstore client initialization is performed regardless of whether the
given replica is the leader or not. This matches the behavior of the k8s
client, and also allows for better health reporting.

One additional difference is that now the kvstore client is always started
if a kvstore backend (i.e., etcd) is configured, rather than additionally
checking whether any feature strictly depending on it is enabled. Still,
this is not expected to make any difference in practice as at least one
feature is practically always expected to be enabled.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: drop the now unused legacy logic

Now that we switched all consumers to the kvstore Cell, let's drop the
legacy initialization logic, as well as the global client accessor.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: rework SetupDummy to directly initialize the client

Now that all users of the LegacyClient global accessor have been
removed, let's also rework the functions setting up the kvstore
client for testing purposes to no longer rely on it.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
ipcache: let watcher/synchronizer directly depend on kvstore client

Now that the kvstore client is provided via Hive, let's update the
ipcache watcher and synchronizers to directly retrieve it.

The watcher is still started inline during the daemon initialization
process, rather than via a job started by the dedicated cell, because
it strictly needs to be started after that the ipcache map has been
recreated by newDaemon.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
status: directly depend on the kvstore client

Remove the usages of the legacy global accessor by directly depending on
the kvstore client.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
nodediscovery: directly depend on the kvstore client

Remove the usages of the legacy global accessor by directly depending on
the kvstore client.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
operator: directly depend on the kvstore client

Remove the usages of the legacy global accessory by directly depending
on the kvstore client. Similarly, let's also drop the access to the
associated global kvstore options, which are also planned to be removed
in a subsequent commit. Again, we slightly change the behavior, starting
the watchdog operations if the kvstore client is configured, but no
feature actually using it is enabled. This is expected to have no impact
in practice, and is also more correct as the cluster config is now
always configured.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
operator/health: directly depend on the kvstore client

Remove the usages of the legacy global accessory by directly depending
on the kvstore client. The check is also updated to let all replicas
report the kvstore status, given that the client is now always started.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore/cell: rework extra options propagation

Let's propagate the extra options by value, rather than pointer, to
allow simplifying the logic and remove a bunch of nil checks, which
are also easy to forget. Let's also mark the field as optional, to
avoid having to provide the options when not necessary, such as in
tests. Finally, let's uniform their configuration, dropping the
misleading setExtraConfig method.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
allocator: directly depend on the kvstore client

Remove the usages of the legacy global accessory by taking the kvstore
client as parameter, and adapt the different users accordingly.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
config: remove the now unused kvstore-related global config options

The DaemonConfig Populate logic depends on whether the kvstore is
enabled or not to perform certain sanity checks, and potentially
modify certain configurations. Although this is not great, let's
keep things simple for the moment and preserve the current behavior.
To this end, we need to retrieve the configuration option directly
via viper, as hive has not been started yet at that point.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
test: provide kvstore client to daemon and controlplane tests

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore/cell: export the Config type

So that other cells can depend on it if they need to retrieve the
parameters, rather than relying on the (to-be-removed) global options.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
restapi: directly depend on the kvstore config

Remove the accesses to the global kvstore options in favor of directly
depending on the kvstore config struct.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
daemon: drop KVstoreEnabled in favor of depending on the client

Remove the KVstoreEnabled function, which relies on the global option,
and adapt the consumers to directly depend on the kvstore client.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
nodediscovery: directly register the kvstore-nodes fence

Directly register the kvstore-nodes fence from the nodediscovery cell,
rather than having it done indirectly by the regenerator one. This also
allows to drop the reference to the to-be-removed global kvstore option
from there.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
allocator: repurpose the kvstore-connectivity-timeout flag

Currently, a subset of the callers of identity allocation functions
configure the context timeout based on the kvstore-connectivity-timeout
flag. Yet, this is suboptimal for two main reasons:

* The name of the flag is misleading, given that it does not control
  any specific kvstore operation, and the allocator can operate both
  in CRD and kvstore mode;
* The usage of the timeout is easy to be missed, and certain callers
  do indeed not specify it.

Let's get these two issues addressed by renaming the flag to
identity-allocation-timeout to make its purpose clear, moving its
definition as part of the identity allocator flags, and letting
the involved functions to directly enforce the timeout internally,
hence removing the burden from the callers.

As a consequence, the kvstore-connectivity-timeout flag is also removed
from the clustermesh-apiserver/kvstoremesh commands, as unused.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: propagate remaining options, rather than accessing the globals

Propagate KVstoreLeaseTTL and KVstoreMaxConsecutiveQuorumErrors to the
kvstore backends as ExtraOptions, to drop the corresponding accesses to
the global option variables.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
store: do not depend on global KVstorePeriodicSync option

Instead, require the callers to explicitly configure this parameter
when relevant. Specifically, we disable periodic synchronization in
the operator, given that no local keys are ever upserted, and
hard-code it in the agents, for simplicity. We use 30 minutes as
interval, considering that the goal is only to automatically recover
from unexpected conditions, and as a trade off to avoid introducing
too many requests in case of very large clusters (the 30 minutes
interval corresponds to ~2.7QPS in a 5K nodes cluster).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: inject backend factory via hive

Slightly refactor the clustermesh logic to inject the factory to create
clients targeting remote clusters via Hive. The rationale is twofold: on
the one hand, it allows dropping the access to the to-be-removed global
kvstore options; on the other, it simplifies testing, allowing to easily
override the client (e.g., in script tests, so that all requests target
the same in-memory representation).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
nodediscovery: propagate context to StartDiscovery and dependants

Propagate the daemon context to the nodediscovery.StartDiscovery(), and
from there to the different API invocations. Most notably, this improves
the shutdown procedure, especially in kvstore mode, aborting ongoing
operations before that the client gets stopped.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
operator: register operator hooks as last cell

Move the registration of the lifecycle hooks responsible for
starting/stopping the operator leader election, and in turn
the cells within the `WithLeaderElection` scope, as last in
the list of operator cells. This is to ensure that the start
hook is always executed last, when all the others outside
the leader election scope are guaranteed to have already
completed, even if not being one of its direct dependencies.
Indeed, any cell inside the leader election scope may depend
on them.

Suggested-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/kvstore/cell.go:
```diff
@@ -4,89 +4,67 @@
+			Opts      ExtraOptions `optional:"true"`
+		}) Client {
+			if in.Config.KVStore == DisabledBackendName {
+				return &clientImpl{enabled: false, cfg: in.Config}
```
Why we return client instance with config here :thinking:? It's a bit confusing to me to propagate config when kvstore is disabled. 

I'm asking to see if we could have two different implementations, one `clientImpl` and for example `noopClientImpl` which would simplify a little bit `clientImpl` implementation (by removing all "if enabled").

giorio94 replied:
> Why we return client instance with config here 🤔? It's a bit confusing to me to propagate config when kvstore is disabled.

The main reason is that a few supposedly kvstore-related options leaked into other subsystems (most notably KVStoreConnectivityTimeout and KVStorePeriodicSync), and are accessed regardless of whether the kvstore is actually enabled or not. KVStoreOpt is also a bit special, because it can be set to tune certain parameters for the clustermesh clients, even though Cilium operates in CRD mode. 

> I'm asking to see if we could have two different implementations, one clientImpl and for example noopClientImpl which would simplify a little bit clientImpl implementation (by removing all "if enabled").

That would be possible, but I'm not sure it would actually simplify things. 

Dropping the `IsEnabled` method is not actually possible, because there's quite a lot of logic here and there in the code-base that only needs to be started if running (or not running) in kvstore mode. 

Besides that, the only two checks of the `enabled` field are in the start/stop hooks, and could be actually dropped because the hooks are currently not registered at all if kvstore support is disabled.

marseel replied:
> The main reason is that a few supposedly kvstore-related options leaked into other subsystems (most notably KVStoreConnectivityTimeout and KVStorePeriodicSync),

yea, that was my point. Looking at the code it seems to me like even though they have "KVStore" name as parameters, they should not belong to KVStore client/config and should belong to identity Allocator Cell or node sync for example, no?
 we could probably call them:
`KVStoreConnectivityTimeout` -> `IdentityConnectivityTimeout`
`KVStorePeriodicSync` -> `IdentitySyncInterval` + `KVStoreNodeSyncInterval` (or even better `NodeSyncInterval` if we add support for CRD as well)

> Dropping the IsEnabled method is not actually possible, because there's quite a lot of logic here and there in the code-base that only needs to be started if running (or not running) in kvstore mode.

Ah, I was thinking about making `clientImpl` implementation skip `if !cl.enabled {` within methods, not removal of `IsEnabled` method. 

> Dropping the IsEnabled method is not actually possible, because there's quite a lot of logic here and there in the code-base that only needs to be started if running (or not running) in kvstore mode.

About `IsEnabled`, In ideal world I would like to treat all the `IsEnabled()` checks spread across code base as optimizations rather then necessity. It's a huge code smell and thinking about it more, I think it would be a very reasonable goal to get rid of it (long-term).
For example, in https://github.com/cilium/cilium/pull/39927/commits/4a8ac8b96a3ee46cf4758ffac7a6217e2cc2b919 I don't see any reason why we are doing these checks only for KVStore and not for CRD. I understand that probably we are waiting on CRD synchronization before, but even if that's true, we should double-check it was synchronized in case someone refactors other parts of code / moves it as it affects correctness of regeneration. 

Anyway, this is interesting discussion and I want to follow-up with some issues after this PR lands, no need to incorporate these comments into it.

giorio94 replied:
>>  The main reason is that a few supposedly kvstore-related options leaked into other subsystems (most notably KVStoreConnectivityTimeout and KVStorePeriodicSync),
>
> yea, that was my point. Looking at the code it seems to me like even though they have "KVStore" name as parameters, they should not belong to KVStore client/config and should belong to identity Allocator Cell or node sync for example, no?
we could probably call them:
KVStoreConnectivityTimeout -> IdentityConnectivityTimeout
KVStorePeriodicSync -> IdentitySyncInterval + KVStoreNodeSyncInterval (or even better NodeSyncInterval if we add support for CRD as well)

Yep, let me have a look at how problematic it is extracting these two flags. That would also avoid adding them to the operator, which is something I definitely don't like of the current approach.

> >    Dropping the IsEnabled method is not actually possible, because there's quite a lot of logic here and there in the code-base that only needs to be started if running (or not running) in kvstore mode.
>
> Ah, I was thinking about making clientImpl implementation skip if !cl.enabled { within methods, not removal of IsEnabled method.

Got it, I had dropped that check in the previous push.

>> Dropping the IsEnabled method is not actually possible, because there's quite a lot of logic here and there in the code-base that only needs to be started if running (or not running) in kvstore mode.
>
> About IsEnabled, In ideal world I would like to treat all the IsEnabled() checks spread across code base as optimizations rather then necessity. It's a huge code smell and thinking about it more, I think it would be a very reasonable goal to get rid of it (long-term).
For example, in https://github.com/cilium/cilium/commit/4a8ac8b96a3ee46cf4758ffac7a6217e2cc2b919 I don't see any reason why we are doing these checks only for KVStore and not for CRD. I understand that probably we are waiting on CRD synchronization before, but even if that's true, we should double-check it was synchronized in case someone refactors other parts of code / moves it as it affects correctness of regeneration.

I totally agree, the two paths should be uniformed as we already had many cases in the past in which we were waiting for synchronization from CRDs but not kvstore, or vice versa. And the fact that we do the operations at different points in time can also lead to tricky situations based on ordering and so on, that can change over time.

That said, as you mentioned, this is definitely something for a follow-up.

giorio94 replied:
> Yep, let me have a look at how problematic it is extracting these two flags. That would also avoid adding them to the operator, which is something I definitely don't like of the current approach.

Pushed the new version, let me know how it looks to you.

marseel commented on pkg/kvstore/client.go:
```diff
@@ -8,52 +8,75 @@ import (
-	if c == nil {
-		err := <-errChan
-		logging.Fatal(scopedLog, "Unable to create kvstore client", logfields.Error, err)
+func (cl *clientImpl) Start(hctx cell.HookContext) (err error) {
```
Just wanted to double-check if I got this right.
Before, we were not blocking on kvstore initialization during bootstrap while we do now and rely on hive ordering to make sure that kvstore is initialized before other dependent components start, right?

giorio94 replied:
Yes, that is the intent. I agree that it modifies the current behavior a bit, but IMO it makes it more straightforward to reason about, and it ensures that any dependent component is guaranteed to be able to access an initialized client at any time. The previous implementation was more fragile instead, because calling `kvstore.Client()` could block if the client had not been initialized yet, and even at that point there was no guarantee that the connection to the kvstore had actually been established successfully for the first time (with the ticking bomb of a subsequent fatal in case of errors). That was mostly needed to handle the historical etcdin-pod-network support, but that's not needed anymore fortunately.

marseel commented on pkg/kvstore/client.go:
```diff
@@ -8,52 +8,75 @@ import (
-	return defaultClient
+func (cl *clientImpl) Stop(cell.HookContext) error {
+	if cl.enabled && cl.BackendOperations != nil {
+		cl.BackendOperations.Close()
```
Before, we were not closing backend on stop:
```
				// We don't explicitly close the backend here, because that would
				// attempt to revoke the lease, causing all entries associated
				// with that lease to be deleted. This would not be the
				// behavior expected by the consumers of this cell.
```
why is it fine to do it now :thinking: ?

giorio94 replied:
Basically because that comment has been out of date for quite some time now. The [current `Close()` implementation](https://github.com/giorio94/cilium/blob/mio%2Fkvstore-cell-2/pkg/kvstore/etcd.go#L1657-L1671) no longer revokes any lease, but only closes the underlying connections and waits for the termination of the lease managers.

marseel commented on daemon/cmd/daemon.go:
```diff
@@ -617,15 +616,6 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-	if option.Config.KVStore == "" {
-		d.logger.Info("Skipping kvstore configuration")
-	} else {
-		bootstrapStats.kvstore.Start()
```
I guess it would be nice to still have the metric for bootstrap (I am not huge fun of this metric anyway TBF)  - but if we are removing it, do we have replacement? I don't think lifecycle hooks have metrics, but that would be a reasonable feature request to hive I think. 
We could at least add some minor release-note saying it's removed.

giorio94 replied:
Yeah, I don't like dropping it either, let me see if there's an easy way to preserve it for the moment.

giorio94 replied:
Updated the commits to preserve the metric. The solution is a bit hacky, but it worked correctly:

```
# HELP cilium_agent_bootstrap_seconds Duration of bootstrap sequence
# TYPE cilium_agent_bootstrap_seconds gauge
cilium_agent_bootstrap_seconds{outcome="success",scope="bpfBase"} 0.000339429
cilium_agent_bootstrap_seconds{outcome="success",scope="cleanup"} 0.000250568
cilium_agent_bootstrap_seconds{outcome="success",scope="daemonInit"} 0.001798756
cilium_agent_bootstrap_seconds{outcome="success",scope="earlyInit"} 0.105324329
cilium_agent_bootstrap_seconds{outcome="success",scope="enableConntrack"} 1.1534e-05
cilium_agent_bootstrap_seconds{outcome="success",scope="fqdn"} 5.8027e-05
cilium_agent_bootstrap_seconds{outcome="success",scope="healthCheck"} 0.000829712
cilium_agent_bootstrap_seconds{outcome="success",scope="ipam"} 0.099911595
cilium_agent_bootstrap_seconds{outcome="success",scope="k8sInit"} 0.250829825
cilium_agent_bootstrap_seconds{outcome="success",scope="kvstore"} 0.003396516
cilium_agent_bootstrap_seconds{outcome="success",scope="mapsInit"} 0.000233739
cilium_agent_bootstrap_seconds{outcome="success",scope="overall"} 6.565432318
cilium_agent_bootstrap_seconds{outcome="success",scope="restore"} 0.003547582
```

marseel commented on pkg/endpoint/regenerator.go:
```diff
@@ -45,26 +47,32 @@ func newRegenerator(in struct {
 
 func (r *Regenerator) WaitForKVStoreSync(ctx context.Context) error {
+	if !r.kvstoreEnabled {
+		return ctx.Err()
```
nit: this seems a bit incorrect to me, let's return nil?
For example, if some usage of that method would print something like:
```
Error waiting on KVStoreSync: context deadline exceeded
```
when someone is not using KVStore, that would be very confusing to me.

giorio94 replied:
Fair enough, let me update it. (I'll wait for a few more reviews before pushing, just to not re-trigger CI).

marseel reviewed (commented):
I did one **very quick** pass to get this review started, I will do one more in depth tomorrow, but looks like an amazing improvement :)

giorio94 committed:
kvstore/cell: enable reporting bootstrap time

In order to preserve updating the bootstrap_seconds metric with
information on the kvstore client setup, let's extend the kvstore
cell to additionally take an optional SpanStat parameter that,
if provided, gets automatically updated with the measurement of
the bootstrap time of the kvstore client.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
allocator: repurpose the kvstore-periodic-sync flag

Let's rename the kvstore-periodic-sync flag to
identity-allocation-sync-interval to make its purpose
clearer, and move it alongside the other allocator
related flags.

As a consequence, the kvstore-periodic-sync flag is also removed
from the clustermesh-apiserver/kvstoremesh commands, as unused.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
golangci-lint: disable exhaustruct for test files

Forcing to configure all fields of Config structs in test files looks
superfluous, and leads to overly verbose tests in certain circumstances.
Hence, let's disable it for tests files only.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

qmonnet reviewed (approved):
OK for changes owned by `docs-structure`, impressive work!

joamaki commented on pkg/kvstore/client.go:
```diff
@@ -20,6 +22,58 @@ var (
 	defaultClientSet = make(chan struct{})
 )
 
+type Client interface {
```
Let's add bit of docs here on what the client is about

joamaki commented on pkg/kvstore/client.go:
```diff
@@ -20,6 +22,58 @@ var (
+
+func (cl *clientImpl) Start(hctx cell.HookContext) (err error) {
+	cl.logger.Info("Establishing connection to kvstore")
+	backend, errCh := NewClient(context.Background(), cl.logger, cl.cfg.KVStore, cl.cfg.KVStoreOpt, cl.opts)
```
Not for this PR, but this background context looks odd given that there's `Close` as well to stop the client. I wonder if we can just get rid of it completely or explicitly only use it for the initial connection attempt? I would maybe also consider making `NewClient` side-effect free and only connect when the client is used and if we want to preserve the behavior of connecting at start we could have e.g. `TryConnect` method or similar.

giorio94 replied:
> Not for this PR, but this background context looks odd given that there's Close as well to stop the client. I wonder if we can just get rid of it completely or explicitly only use it for the initial connection attempt?

Yeah, makes sense. It should be doable having it control only the initial setup, even though it may be a bit of work to make sure it does not break anything

> I would maybe also consider making NewClient side-effect free and only connect when the client is used and if we want to preserve the behavior of connecting at start we could have e.g. TryConnect method or similar.

I personally prefer having the connection being explicitly initialized by the start hook, to basically make sure that it can succeed at that point in time. That's obviously not a guarantee that things will not break later, but it ensures that we can catch the vast majority of the problems early in the Cilium agent setup, and not in the middle of the initialization. That used to be a problem previously, as you could have a random call just block because the client had not connected yet, and these are significantly harder to reason about IMO. 

That said, I agree that the `NewClient` name is definitely unfortunate, considering it actual behavior. One possibility could be simply renaming it; another could be extracting a `Start` method that's called to actually start the connection procedure.

joamaki commented on pkg/kvstore/cell.go:
```diff
@@ -19,32 +19,39 @@ import (
 const DisabledBackendName = ""
 
 // Cell returns a cell which provides a promise for the global kvstore client.
```
Outdated comment

joamaki commented on operator/cmd/root.go:
```diff
@@ -82,6 +82,10 @@ var (
 		Infrastructure,
 		ControlPlane,
+
+		cell.Invoke(
```
Let's add a comment here that this needs to be last so that everything outside leader lifecycle is started before the leader lifecycle.

joamaki reviewed (approved):
Awesome!

marseel commented on pkg/identity/cache/allocator.go:
```diff
@@ -96,13 +96,25 @@ type CachingIdentityAllocator struct {
 	// performed before failing. This is mainly introduced for testing purposes.
 	maxAllocAttempts int
+
+	// timeout for identity allocation operations.
```
nit: I would remove "allocation" as it also is used for release/lookup etc.

giorio94 replied:
Yeah, I'm referring to _allocation_ in wide sense here, considering all related operations. The flag is also named `identity-allocation-timeout` to make it a bit more specific, as just `identity-timeout` seemed a bit ambiguous to me.

marseel reviewed (approved):
Thanks Marco for the amazing work.
general nit: I would rename all places where we call kvstore.Client as backend and switch it to client/kvstoreClient, it's very inconsistent now. 

Based on the review I think there are some followup not directly related to this PR, but generally to the code that this PR was somehow touching:

- Block UpdateCiliumNodeResource after starting discovery / codify this comment: https://github.com/cilium/cilium/blob/49c2c6c9861a487541e76a15a4e0f532868ea9aa/pkg/nodediscovery/nodediscovery.go#L204-L207
- Remove SharedStore Close method - only used in testing
- Rename DeleteLocalKey to DeleteKey - it actually deletes non-local keys
- Investigate if we really need to fetch from kvstore here: https://github.com/cilium/cilium/blob/160b9c8cbe66f99d299bc4a9e8457fcc25b04f1b/pkg/identity/cache/cache.go#L271-L274
It seems that it can return nil identity when having connectivity problems with external kvstore even if identity exists - could lead to very interesting inconsistent behaviors
- BootstrapTime metrics rework to allow for easier injection and change from reporting duration to timestamp when particular action finished

Some other ideas:
- Unify kvstore/CRD identity allocation and/or remove slaveKeys from kvstore implementation
- Get rid of some of kvstore.IsEnabled checks - It seems like there are in places where they shouldn't be, for example https://github.com/cilium/cilium/blob/14c4398033caf05c8517d029c5472822d3324891/pkg/endpoint/regenerator.go#L71 (as regeneration should wait for corresponding information regardless of backend) or https://github.com/cilium/cilium/blob/c2ed91b54f245912b2b6483a5f5b3aad7b96d283/daemon/cmd/daemon_main.go#L1597 (similar concern)

giorio94 committed:
operator/service-sync: directly depend on the kvstore client

Now that kvstore.Client is also provided in the operator, let's rework
the service-sync cell to directly depend on it. Additionally, let's
simplify the store initialization, given that it is guaranteed to not
be blocking.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

derailed reviewed (approved):
@giorio94 Dang! that is a whopper... Nice work Marco!

tklauser reviewed (approved):
LGTM for @cilium/fqdn
